mempool/src/pool/reorg.rs

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Support for updating the mempool upon a reorg
+
+use std::{collections::BTreeSet, mem};
+
+use chainstate::chainstate_interface::ChainstateInterface;
+use common::{
+    chain::{Block, GenBlock, SignedTransaction},
+    primitives::{Id, Idable},
+};
+use logging::log;
+use utils::{shallow_clone::ShallowClone, tap_error_log::LogError};
+use utxo::UtxosStorageRead;
+
+use crate::{
+    get_memory_usage::GetMemoryUsage,
+    pool::{tx_verifier, Mempool, MempoolStore},
+};
+
+/// An error that can happen in mempool on chain reorg
+#[derive(Debug, thiserror::Error, PartialEq, Eq)]
+pub enum ReorgError {
+    #[error(transparent)]
+    Chainstate(#[from] chainstate::ChainstateError),
+    #[error("Could not find the previous tip")]
+    OldTip,
+    #[error("Could not find the previous tip index")]
+    OldTipIndex,
+    #[error("Could not find the new tip index")]
+    NewTipIndex,
+    #[error("Block {0:?} not found while traversing history")]
+    BlockNotFound(Id<Block>),
+    #[error("Chainstate call: {0}")]
+    ChainstateCall(#[from] subsystem::subsystem::CallError),
+}
+
+/// Iterate over blocks following the parent link. Apply given function to each block.
+fn for_each_block(
+    chainstate: &dyn ChainstateInterface,
```

Comment:
Please avoid dyn when possible. Let's just use impl ChainstateInterface (or generic explicitly).

Reply:
Yea, I had impl previously but then it's a private function that is only used once with `&dyn ChainstateInterface` so I thought it's a bit excessive to generalise it. Can put the `impl` back, not a big deal